File: android_app/src/main/java/app/time_to/timeto/ui/FolderFormSheet.kt
package app.time_to.timeto.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import app.time_to.timeto.rememberVM
import app.time_to.timeto.setFalse
import timeto.shared.db.TaskFolderModel
import timeto.shared.vm.FolderFormSheetVM

@Composable
fun FolderFormSheet(
    isPresented: MutableState<Boolean>,
    folder: TaskFolderModel?,
) {
    TimetoSheet(isPresented = isPresented) {

        val (vm, state) = rememberVM(folder) { FolderFormSheetVM(folder) }

        Column(
            modifier = Modifier
                .fillMaxHeight()
                .background(c.bgFormSheet)
        ) {

            val scrollState = rememberScrollState()

            SheetHeaderView(
                onCancel = { isPresented.value = false },
                title = state.headerTitle,
                doneText = state.headerDoneText,
                isDoneEnabled = state.isHeaderDoneEnabled,
                scrollToHeader = scrollState.value,
            ) {
                vm.save {
                    isPresented.value = false
                }
            }

            Column(
                modifier = Modifier
                    .verticalScroll(
                        state = scrollState
                    )
                    .padding(bottom = 20.dp)
                    .navigationBarsPadding()
                    .imePadding()
            ) {

                MyListView__Padding__SectionHeader()

                MyListView__HeaderView(
                    state.inputNameHeader,
                )

                MyListView__Padding__HeaderSection()

                MyListView__ItemView(
                    isFirst = true,
                    isLast = true,
                ) {
                    MyListView__ItemView__TextInputView(
                        placeholder = state.inputNamePlaceholder,
                        text = state.inputNameValue,
                        onTextChanged = { newText -> vm.setInputNameValue(newText) },
                    )
                }

                if (folder != null && !folder.isToday) {

                    MyListView__Padding__SectionSection()

                    MyListView__ItemView(
                        isFirst = true,
                        isLast = true,
                    ) {
                        MyListView__ItemView__ActionView(
                            text = state.deleteFolderText,
                        ) {
                            vm.delete {
                                isPresented.setFalse()
                            }
                        }
                    }
                }
            }
        }
    }
}

File: ios_app/iphone/FolderFormSheet.swift
import SwiftUI
import shared

struct FolderFormSheet: View {

    @State private var vm: FolderFormSheetVm
    @Binding private var isPresented: Bool
    @State private var sheetHeaderScroll = 0

    init(
            isPresented: Binding<Bool>,
            folder: TaskFolderDb?
    ) {
        _isPresented = isPresented
        vm = FolderFormSheetVm(folder: folder)
    }

    var body: some View {

        VMView(vm: vm, stack: .VStack(spacing: 0)) { state in

            SheetHeaderView(
                    onCancel: { isPresented.toggle() },
                    title: state.headerTitle,
                    doneText: state.headerDoneText,
                    isDoneEnabled: state.isHeaderDoneEnabled,
                    scrollToHeader: sheetHeaderScroll
            ) {
                vm.save {
                    isPresented = false
                }
            }

            ScrollViewWithVListener(showsIndicators: false, vScroll: $sheetHeaderScroll) {

                VStack(spacing: 0) {

                    VStack(spacing: 0) {

                        MyListView__HeaderView(title: state.inputNameHeader)
                                .padding(.top, 20)

                        MyListView__Padding__HeaderSection()

                        MyListView__ItemView(
                                isFirst: true,
                                isLast: true
                        ) {

                            MyListView__ItemView__TextInputView(
                                    text: state.inputNameValue,
                                    placeholder: state.inputNamePlaceholder,
                                    isAutofocus: false,
                                    onValueChanged: { newValue in
                                        vm.setInputNameValue(name: newValue)
                                    }
                            )
                        }

                        if let folder = vm.folder, !folder.isToday {

                            MyListView__Padding__SectionSection()

                            MyListView__ItemView(
                                    isFirst: true,
                                    isLast: true
                            ) {

                                MyListView__ItemView__ActionView(
                                        text: state.deleteFolderText
                                ) {
                                    vm.delete {
                                        isPresented = false
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
                .background(c.sheetBg)
    }
}
